### PR TITLE
Consider the scope of multiValueHeaders and multiValueQueryStringParameters

### LINKED ISSUE #1486: Support for multi-value parameters in Amazon API Gateway and ELB to Lambda functions
## Is your feature request related to a problem? Please describe.

At flaconi we are using heavily nodejs lambdas which are coming through an elb, these lambdas are using multi Value parameters, `multiValueQueryStringParameters ` and `multiValueHeaders `, as this sdk is only looking for the normal headers field, no trace header is parsed and we would be forced to manually instrument these which is too much of a work. Also sind end of 2018 api gateway is supporting multi value parameters. 

## Feature Description
In the check and parsing of [the event in the sdk](https://github.com/newrelic/node-newrelic/blob/22958a0c53e8faa35eb45d32cbe82acedede486b/lib/serverless/api-gateway.js#L45) we need to enrich it to also check for multi value parameters, evtl other parts need also be touched, that is the one where I saw why it does not work for our lambdas

## Priority
Blocker

Also I'm happy to provide an pr for this feature request if you would like

### PR DESCRIPTION
**Pull request by Engerim**
## Proposed Release Notes

* Add support for [Multi Value Parameters](https://aws.amazon.com/blogs/compute/support-for-multi-value-parameters-in-amazon-api-gateway) from API Gateway and ALB events for Lambdas (Thank you @Engerim for your contribution 🙏🏻)

## Links
* https://aws.amazon.com/de/blogs/compute/support-for-multi-value-parameters-in-amazon-api-gateway/
* https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#multi-value-headers
* Closes #1486 

## Details
The LambdaProxyWebRequest and LambdaProxyWebResponse classes where adjusted to handle multi value parameters.
Also the transport type and port are now retrieved from a lowercased headers map to handle graceful difference between api gateway events and alb event, (alb event headers are lowercased by default and api gateway event headers not). Also the test if it is a lambda proxy event has been adjusted through this tracing for alb lambdas with multi value headers is possible. 
 
* enrich the LambdaProxyWebRequest and LambdaProxyWebResponse class to handle events where multi value parameters are set, like events for alb or api gateway with multi value header support
* add test case for alb event with multi value header

### CODE REVIEW

**Comment by Engerim** [comment]
@bizob2828  here is the pr, I have tested it in our infrastructure with our alb lambdas and tracing worked finally and also the transaction tracking, as it is made in a backward compatibility way, it should not harm any lambdas with events with only normal header and query string parameters.

**Comment by jmartin4563** [comment]
Hiya @Engerim! Taking a look at this, thank you for the submission! Just a fyi that you'll need to rebase to get the versioned tests passing for this PR, there was an unrelated issue that was addressed in #1490

**Comment by Engerim** [comment]
@jmartin4563 I have updated the branch should be now fine to pass

**Review by jmartin4563 — CHANGES_REQUESTED** [review]
Still want to do a test with a lambda on my end, but have a couple of thoughts just going over the changes, let me know what you think!

File: lib/serverless/api-gateway.js [hunk 1]
```diff
@@ -30,11 +32,53 @@ class LambdaProxyWebRequest {
...
 class LambdaProxyWebResponse {
   constructor(lambdaResponse) {
-    this.headers = lambdaResponse.headers
+    this.headers = normalizeHeaders(lambdaResponse)
     this.statusCode = lambdaResponse.statusCode
   }
 }
 
+/**
+ * normalizes headers either from multi value headers or normal headers to a
```
[the remark this answers is not in the document]
Comment (Engerim) [reply lib/serverless/api-gateway.js:41]: sure was a copy & paste issue

File: test/unit/serverless/lambda-sample-events.js [hunk 2]
```diff
@@ -489,6 +489,40 @@ const sesEvent = {
...
+    'upgrade-insecure-requests': ['1'],
+    'user-agent': [
+      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36'
+    ],
+    'x-amzn-trace-id': ['Root=1-5c536348-3d683b8b04734faae651f476'],
+    'x-forwarded-for': ['72.12.164.125'],
+    'x-forwarded-port': ['80'],
+    'x-forwarded-proto': ['http'],
+    'x-imforwards': ['20']
+  },
```
[the remark this answers is not in the document]
Comment (Engerim) [reply test/unit/serverless/lambda-sample-events.js:521]: I can add one header with multiple values and then will test that these values are also reflected in the transaction

File: lib/serverless/api-gateway.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Engerim) [reply lib/serverless/api-gateway.js]: is adjsuted

File: test/unit/serverless/lambda-sample-events.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Engerim) [reply test/unit/serverless/lambda-sample-events.js:525]: added a set-cookie header with multiple values and extended the test case to check for this value, also added a check for the multi query string parameter

File: lib/serverless/api-gateway.js [hunk 3]
```diff
@@ -30,11 +32,53 @@ class LambdaProxyWebRequest {
...
+ * lowercase key map with comma separated string
+ *
+ * @param {object} event The event with headers to normalize
+ * @param {boolean} lowerCaseKey Whether to lowercase the header names or not
+ * @returns {Object<string, string>} The normalized headers map
+ */
+function normalizeHeaders(event, lowerCaseKey = false) {
+  return Object.fromEntries(
+    Object.entries(event.multiValueHeaders ?? event.headers).map(([headerKey, headerValue]) => {
```
Comment (jmartin4563) [inline lib/serverless/api-gateway.js:71]: From reading the AWS docs/testing, it appears that both `multiValueHeaders` and `headers` can be defined at the same time, so I think we'll need to update this function to try and normalize both keys.

**Review by jmartin4563 — CHANGES_REQUESTED** [review]

File: lib/serverless/api-gateway.js (hunk 3 above)
Comment (Engerim) [reply lib/serverless/api-gateway.js:71]: for api gateway yes but for alb not, also if you have multiValueHeaders, all the values are the same as in the normal headers key, so no value in it to normalize it also when the multi value headers are send. the output will be even less. 

also what should happen for a header with multiple values and the same header with only a single value

Comment (Engerim) [reply lib/serverless/api-gateway.js:71]: Also in the context of a lambda which is moved through the application load balancer, you only have mutliValueHeaders or normal headers but never both.

**Review by jmartin4563 — APPROVED** [review]
Thank you again for the submission and your patience @Engerim 🎉

**Comment by Engerim** [comment]
thx a lot for merging this

**Comment by jmartin4563** [comment]
And just a fyi that we plan on doing a release today (most likely) or tomorrow at the latest, so this will go out as part of `9.10.0`

**Pull request merged**